doc/source/ray-air/examples/pytorch_resnet_finetune.ipynb

```diff
@@ -0,0 +1,1152 @@
+    "\n",
+    "            running_loss = 0.0\n",
+    "            running_corrects = 0\n",
+    "\n",
+    "            # Create a dataset iterator for the shard on the current worker\n",
+    "            dataset_iterator = datasets[phase].iter_torch_batches(\n",
+    "                batch_size=worker_batch_size, device=device\n",
+    "            )\n",
+    "            for batch in dataset_iterator:\n",
+    "                inputs = batch[\"image\"]\n",
+    "                labels = batch[\"label\"]\n",
+    "\n",
+    "                # zero the parameter gradients\n",
+    "                optimizer.zero_grad()\n",
+    "\n",
+    "                # forward\n",
+    "                with torch.set_grad_enabled(phase == \"train\"):\n",
+    "                    # Get model outputs and calculate loss\n",
+    "                    outputs = model(inputs)\n",
+    "                    loss = criterion(outputs, labels)\n",
+    "\n",
+    "                    # backward + optimize only if in training phase\n",
+    "                    if phase == \"train\":\n",
+    "                        loss.backward()\n",
+    "                        optimizer.step()\n",
+    "\n",
+    "                # calculate statistics\n",
+    "                running_loss += loss.item() * inputs.size(0)\n",
+    "                running_corrects += evaluate(outputs, labels)\n",
+    "\n",
+    "            epoch_loss = running_loss / datasets[phase].count()\n",
+    "            epoch_acc = running_corrects / datasets[phase].count()\n",
+    "\n",
+    "            if session.get_world_rank() == 0:\n",
+    "                print(\n",
+    "                    \"Epoch {}-{} Loss: {:.4f} Acc: {:.4f}\".format(\n",
+    "                        epoch, phase, epoch_loss, epoch_acc\n",
+    "                    )\n",
+    "                )\n",
+    "\n",
+    "            # Report metrics and checkpoint every epoch\n",
+    "            if phase == \"val\":\n",
+    "                checkpoint = TorchCheckpoint.from_dict(\n",
+    "                    {\n",
+    "                        \"epoch\": epoch,\n",
+    "                        \"model\": model.module.state_dict(),\n",
+    "                        \"optimizer_state_dict\": optimizer.state_dict(),\n",
+    "                    }\n",
+    "                )\n",
+    "                session.report(\n",
+    "                    metrics={\"loss\": epoch_loss, \"acc\": epoch_acc},\n",
+    "                    checkpoint=checkpoint,\n",
+    "                )\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "VmZQKtWnrLT3"
+   },
+   "source": [
+    "Next, setup the TorchTrainer:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {
+    "cellView": "form",
+    "id": "6jeTo6YTrLT3",
+    "tags": [
+     "remove-cell"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "# @title +\n",
+    "if SMOKE_TEST:\n",
```

Comment:
can you explain what is going on with all the s3 stuff?

Reply:
is there a way for me to run this notebook without having to setup s3? (for example running it directly on google colab)

Reply:
Basically this is a mock s3 server, which opens a port on localhost and save the files on local disk. The intention here is to encourage the users to use cloud storage. 

Also, we put these S3 bucket things in hidden cells, users will not see them in our doc. And you can directly run this notebook in colab.